wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -575,67 +583,85 @@ public Instance getMachine(String workspaceId, String machineId) throws NotFound
     }
 
     /**
-     * Removes all states from the in-memory storage, while
-     * {@link CheEnvironmentEngine} is responsible for environment destroying.
+     * Returns true if there is at least one workspace running(it's status is
+     * different from {@link WorkspaceStatus#STOPPED}), otherwise returns false.
      */
-    @PreDestroy
-    @VisibleForTesting
-    void cleanup() {
-        isPreDestroyInvoked = true;
+    public boolean isAnyRunning() {
+        return !states.isEmpty();
+    }
 
-        // wait existing tasks to complete
-        sharedPool.terminateAndWait();
+    /**
+     * Once called no more workspaces are allowed to start, {@link #startAsync}
+     * will always throw an appropriate exception. All the running workspaces
+     * will continue running, unless stopped directly.
+     *
+     * @return true if this is the caller is the one who refused start,
+     * otherwise if start is being already refused returns false
+     */
+    public boolean refuseWorkspacesStart() {
```

Comment:
Workspace manager knows about snapshot creation strategy, so it MUST stop running workspaces to snapshot them before(if it's configured to do so). As workspace runtimes state is managed by this component we need middle state that allows us to handle running workspace and rejects starts of new ones, but stop/snapshot through the API is still allowed. On the other hand component shutdown cleans in memory storage and doesn't allow to make any modification operations in locks, so in this way we guarantee that threads that still working with some in memory data won't continue working and will stop immediately as they meet `isShutdown = true` in the lock. 
So starting thread that wants to change its status from STARTING -> RUNNING is either interrupted by workspace manager or failed because of service shutdown, or everything is okay and workspace status changed, but there is no way to get null workspace or workspace in inconsistent state while it has to be STARTING.